### ISSUE TITLE
[feature request] Context.parseSingleExpr

### ISSUE THREAD

**Issue by back2dos** [HaxeFoundation/haxe#6479]
#6475 made me think of this again. I would like a function like this:

```haxe
static function parseSingleExpr(s:String, pos:Position):Expr;
```

The difference is that not the whole string is expected to be an expression, but only so much is consumed, until the expression is terminated (essentially what hscript's parseExpr does as opposed to parseFullExpr). So I can call it on e.g. `a.b)blargh` and it will return `a.b` while I can use the `pos.max - pos.min` of the returned expression to know how much of the string was not consumed (or we could return `{ expr: Expr, rest:String }`). It would greatly facilitate picking out Haxe code embedded into other languages.

**Comment by Simn** [comment]
Hmm, I like the idea, but I'm not sure it makes for a great API. Will have to think about it a bit.

**Comment by back2dos** [comment]
Yeah. On second thought, returning `{ expr: Expr, rest:String }` makes it far more evident what the function does. Let's pretend I proposed that ;)

**Comment by nadako** [comment]
I created a branch with this function https://github.com/HaxeFoundation/haxe/tree/parse_single_expr, check if that helps with your dsl magic :)

**Comment by back2dos** [comment]
Thanks a lot. Will try my luck with it later this week ;)

**Comment by jcward** [comment]
Wait a second, @back2dos, expressions are recursive, so how useful is "parse _one_ expression" -- that could consume a bunch of the string, with the custom tokens buried somewhere deep inside the "one" expression, couldn't it? At the time it dies on the custom tokens, who knows how many expressions are still outstanding up the stack, right?

Wouldn't it make more sense (from the perspective of the macro author, not necessarily the compiler team ❓  :smile:) to provide:

```
Context.customParseString(codeString:String, customParser:String->Int->{ e:Expr, len:Int }, pos:Position);
```

Where customParser is not a fallback in case of Haxe syntax errors, but an override. That is, I get a chance to consume any tokens I want _before_ Haxe does.

e.g. for consuming XML:

```
function customJSXParser(codeString:String, pos:Int):{ e:Expr, length_consumed:Int } {
  // consume whitespace? pos++
  if (codeString.charAt(pos)!='<') return null; // Performance: quick exit
  // More fancy checking, etc for <tag ... ugly code counting tags .../tag>
  if (not found) return null;
  var rtn:Expr = MyMacro.process_xml( identified_substring );
  return { e:rtn, length_consumed:identified_substring.length };
}
```

This way, all the ugly tag counting is in the user's macro code, and once I'm in a macro, I can literally consume any syntax I want. It's reentrant+recursive because MyMacro.process_xml calls customParseString.

So in my mind it works... my only questions are... 1) is it technically possible? and 2) do you want to open up syntax like that? How ugly will things get? To that, I might argue that it's uglier for users to try and abuse legal Haxe syntax to achieve their DSL desires... Ahem @Simn : http://try-haxe.mrcdk.com/#2e123 😜

**Comment by jcward** [comment]
FYI, I played with `Context.parseSingleExpr` -- here's [the test](https://gist.github.com/jcward/580e4012aed175b799956424f7aa7629) (results in comments). Perhaps it's not working as Juraj envisioned, perhaps it has bugs, or perhaps I don't understand. But I'm not sure how that would help me parse a string of mixed Haxe and other syntax...

Another random idea -- HScript could help you determine where in your string the valid Haxe stops and the custom syntax starts... but of course, it has its limitations (Then again, since you aren't interested in eval, simply parsing, maybe you could fill in some of those missing tokens...) I bet it'd be pretty easy to implement my lex_override idea in HScript... Then with HScript, you could perhaps inject your string with proper delimiters (or wrap in macro calls -- a heredoc syntac would help avoid quote escaping... :wink:) before passing the now-valid Haxe syntax string to the actual `Context.ParseInlineString` ...

**Comment by back2dos** [comment]
Yeah, hscript would work. Right now I'm brace counting in tink_hxx and in practice it's good enough. But reimplementing parsing logic that is in the compiler on the macro interpreter just seems a bit silly to me, hence my proposal to expose it.

The question of reentrancy/recursion with inline markup is a different one. If balanced `<tag ... </tag>` did result in `EMarkup`, then `Context.parseSingleExpr` would just return something with (at least potentially) an `EMarkup` in there. The caller would have to decide what to do with that, which to me is just fine.

**Comment by Simn** [comment]
Let's add this for 4.2, I think it's quite useful and should be easy to implement.
